improver_tests/threshold/test_ThresholdPytest.py

```diff
@@ -0,0 +1,599 @@
+    (vicinity) returns the expected values when a mask is provided
+    to limit the spread of values to matching types, e.g. a high
+    probability on land is only spread to other land points.
+    """
+    plugin = Threshold(**kwargs)
+    result = plugin(custom_cube, landmask)
+
+    assert result.data.shape == expected_result.shape
+    assert np.allclose(result.data, expected_result)
+    assert type(result.data) == type(expected_result)
+    assert result.data.dtype == expected_result.dtype
+
+
+@pytest.mark.parametrize("n_realizations,data", [(1, np.array([[0, np.nan], [1, 1]]))])
+def test_nan_handling(custom_cube):
+    """Test that an exception is raised if the input data contains an
+    unmasked NaN."""
+
+    plugin = Threshold(threshold_values=0.5)
+    with pytest.raises(ValueError, match="NaN detected in input cube data"):
+        plugin(custom_cube)
+
+
+@pytest.mark.parametrize(
+    "n_realizations,data,mask", [(1, np.zeros((2, 2)), np.zeros((2, 2)))]
+)
+def test_landmask_no_vicinity(custom_cube, landmask):
+    """Test that an exception is raised if a landmask is provided but
+    vicinity processing is not being applied."""
+
+    plugin = Threshold(threshold_values=0.5)
+    msg = "Cannot apply land-mask cube without in-vicinity processing"
+    with pytest.raises(ValueError, match=msg):
+        plugin(custom_cube, landmask)
+
+
+@pytest.mark.parametrize("n_realizations,data", [(1, np.zeros((2, 2)))])
+def test_cell_methods(custom_cube):
+    """Test that cell methods are modified as expected when present
+    on the input cube."""
+
+    custom_cube.add_cell_method(CellMethod("max", coords="time"))
+    plugin = Threshold(threshold_values=0.5)
+    result = plugin(custom_cube)
+
+    (cell_method,) = result.cell_methods
+    assert cell_method.method == "max"
+    assert cell_method.coord_names[0] == "time"
+    assert cell_method.comments[0] == "of precipitation_rate"
+
+
+@pytest.mark.parametrize("n_realizations,data", [(4, np.zeros((4, 2, 2)))])
+def test_percentile_collapse(custom_cube):
+    """Test that a percentile coordinate can be collapsed to calculate
+    an average. These percentiles must be equally spaced and centred on
+    the 50th percentile ensure they may be considered as equivalent to
+    realizations."""
+
+    expected_result = np.zeros((2, 2, 2), dtype=np.float32)
+    expected_result[0, 0, 0] = 0.25
+
+    percentile_coord = DimCoord([20, 40, 60, 80], long_name="percentile", units="%")
+    custom_cube.remove_coord("realization")
+    custom_cube.add_dim_coord(percentile_coord, 0)
+    custom_cube.data[0, 0, 0] = 1
+
+    plugin = Threshold(threshold_values=[0.5, 1.5], collapse_coord="percentile")
+    result = plugin(custom_cube)
+
+    assert not result.coords("percentile")
+    assert np.allclose(result.data, expected_result)
+    assert (result.coord(var_name="threshold").points == [0.5, 1.5]).all()
+
+
+def test_threshold_unit_conversion(default_cube):
```

Comment:
I was expecting to see `assert result.coord(var_name="threshold").units == "mm hr-1")